<launch>
  <arg name="map_size_x_"/>
  <arg name="map_size_y_"/>
  <arg name="map_size_z_"/>

  <arg name="odometry_topic"/>
  <arg name="camera_pose_topic"/>
  <arg name="depth_topic"/>
  <arg name="cloud_topic"/>

  <arg name="cx"/>
  <arg name="cy"/>
  <arg name="fx"/>
  <arg name="fy"/>

  <arg name="max_vel"/>
  <arg name="max_acc"/>
  <arg name="planning_horizon"/>

  <arg name="point_num"/>
  <arg name="point0_x"/>
  <arg name="point0_y"/>
  <arg name="point0_z"/>
  <arg name="point1_x"/>
  <arg name="point1_y"/>
  <arg name="point1_z"/>
  <arg name="point2_x"/>
  <arg name="point2_y"/>
  <arg name="point2_z"/>
  <arg name="point3_x"/>
  <arg name="point3_y"/>
  <arg name="point3_z"/>
  <arg name="point4_x"/>
  <arg name="point4_y"/>
  <arg name="point4_z"/>

  <arg name="flight_type"/>
  <arg name="use_distinctive_trajs"/>

  <arg name="obj_num_set"/>

  <arg name="drone_id"/>


  <!-- main node -->
  <!-- <node pkg="ego_planner" name="ego_planner_node" type="ego_planner_node" output="screen" launch-prefix="valgrind"> -->
  <node pkg="ego_planner" name="drone_$(arg drone_id)_ego_planner_node" type="ego_planner_node" output="screen" >
  
    <remap from="~odom_world" to="$(arg odometry_topic)"/>
    <remap from="~planning/bspline" to = "/drone_$(arg drone_id)_planning/bspline"/>
    <remap from="~planning/data_display" to = "/drone_$(arg drone_id)_planning/data_display"/>
    <remap from="~planning/broadcast_bspline_from_planner" to = "/broadcast_bspline"/>
    <remap from="~planning/broadcast_bspline_to_planner" to = "/broadcast_bspline"/>

    <!-- <remap from="~grid_map/odom" to="/uav_$(arg drone_id)/odometry_sensor1/odometry"/>
    <remap from="~grid_map/cloud" to="/drone_$(arg drone_id)_$(arg cloud_topic)"/>
    <remap from="~grid_map/pose"   to = "/drone_$(arg drone_id)_$(arg camera_pose_topic)"/> 
    <remap from="~grid_map/depth" to = "/uav_$(arg drone_id)/vi_sensor/camera_depth/depth/disparity"/> -->
    

    <!-- planning fsm -->
    <param name="fsm/flight_type" value="$(arg flight_type)" type="int"/>
    <param name="fsm/thresh_replan_time" value="1.0" type="double"/>
    <param name="fsm/thresh_no_replan_meter" value="1.0" type="double"/>
    <param name="fsm/planning_horizon" value="9" type="double"/> <!--always set to 1.5 times grater than sensing horizen7.5-->
    <param name="fsm/planning_horizen_time" value="3" type="double"/>
    <param name="fsm/emergency_time" value="1.0" type="double"/>
    <param name="fsm/realworld_experiment" value="false"/>
    <param name="fsm/fail_safe" value="true"/>

    <param name="fsm/waypoint_num" value="$(arg point_num)" type="int"/>
    <param name="fsm/waypoint0_x" value="$(arg point0_x)" type="double"/>
    <param name="fsm/waypoint0_y" value="$(arg point0_y)" type="double"/>
    <param name="fsm/waypoint0_z" value="$(arg point0_z)" type="double"/>
    <param name="fsm/waypoint1_x" value="$(arg point1_x)" type="double"/>
    <param name="fsm/waypoint1_y" value="$(arg point1_y)" type="double"/>
    <param name="fsm/waypoint1_z" value="$(arg point1_z)" type="double"/>
    <param name="fsm/waypoint2_x" value="$(arg point2_x)" type="double"/>
    <param name="fsm/waypoint2_y" value="$(arg point2_y)" type="double"/>
    <param name="fsm/waypoint2_z" value="$(arg point2_z)" type="double"/>
    <param name="fsm/waypoint3_x" value="$(arg point3_x)" type="double"/>
    <param name="fsm/waypoint3_y" value="$(arg point3_y)" type="double"/>
    <param name="fsm/waypoint3_z" value="$(arg point3_z)" type="double"/>
    <param name="fsm/waypoint4_x" value="$(arg point4_x)" type="double"/>
    <param name="fsm/waypoint4_y" value="$(arg point4_y)" type="double"/>
    <param name="fsm/waypoint4_z" value="$(arg point4_z)" type="double"/>

    <param name="grid_map/resolution"      value="0.1" /> 
    <param name="grid_map/map_size_x"   value="$(arg map_size_x_)" /> 
    <param name="grid_map/map_size_y"   value="$(arg map_size_y_)" /> 
    <param name="grid_map/map_size_z"   value="$(arg map_size_z_)" /> 
    <param name="grid_map/local_update_range_x"  value="5.5" /> 
    <param name="grid_map/local_update_range_y"  value="5.5" /> 
    <param name="grid_map/local_update_range_z"  value="4.5" /> 
    <param name="grid_map/obstacles_inflation"     value="0.15" /> 
    <param name="grid_map/local_map_margin" value="10"/>
    <param name="grid_map/ground_height"        value="-0.01"/>
    <!-- camera parameter -->
    <param name="grid_map/cx" value="$(arg cx)"/>
    <param name="grid_map/cy" value="$(arg cy)"/>
    <param name="grid_map/fx" value="$(arg fx)"/>
    <param name="grid_map/fy" value="$(arg fy)"/>
    <!-- depth filter -->
    <param name="grid_map/use_depth_filter" value="true"/>
    <param name="grid_map/depth_filter_tolerance" value="0.15"/>
    <param name="grid_map/depth_filter_maxdist"   value="5.0"/>
    <param name="grid_map/depth_filter_mindist"   value="0.2"/>
    <param name="grid_map/depth_filter_margin"    value="2"/>
    <param name="grid_map/k_depth_scaling_factor" value="1000.0"/>
    <param name="grid_map/skip_pixel" value="2"/>
    <!-- local fusion -->
    <param name="grid_map/p_hit"  value="0.65"/>
    <param name="grid_map/p_miss" value="0.35"/>
    <param name="grid_map/p_min"  value="0.12"/>
    <param name="grid_map/p_max"  value="0.90"/>
    <param name="grid_map/p_occ"  value="0.80"/>
    <param name="grid_map/min_ray_length" value="0.1"/>
    <param name="grid_map/max_ray_length" value="4.5"/>

    <param name="grid_map/virtual_ceil_height"   value="3.0"/>
    <param name="grid_map/visualization_truncate_height"   value="2.9"/>
    <param name="grid_map/show_occ_time"  value="false"/>
    <param name="grid_map/pose_type"     value="2"/>  
    <param name="grid_map/frame_id"      value="world"/>






    <remap from="/global_cloud" to="/$(arg cloud_topic)"/>
    <remap from="/odom_topic" to="/$(arg odometry_topic)"/>
    <remap from="/depth_topic" to="/pcl_render_node/depth" />
    
    <param name="occ_map/origin_x" value="-25.0" type="double"/>
    <param name="occ_map/origin_y" value="-25.0" type="double"/>
    <param name="occ_map/origin_z" value="0.0" type="double"/>
    <param name="occ_map/map_size_x" value="$(arg map_size_x_)" type="double"/>
    <param name="occ_map/map_size_y" value="$(arg map_size_y_)" type="double"/>
    <param name="occ_map/map_size_z" value="$(arg map_size_z_)" type="double"/>
    <param name="occ_map/local_radius_x"  value="5.5" /> 
    <param name="occ_map/local_radius_y"  value="5.5" /> 
    <param name="occ_map/local_radius_z"  value="4.5" /> 
    <param name="occ_map/resolution" value="0.1" type="double"/>
    <param name="occ_map/show_raw_depth" value="true" type="bool"/>
    <param name="occ_map/show_filter_proj_depth" value="true" type="bool"/>
		<param name="occ_map/depth_scale" value="1000.0"/>
		<param name="occ_map/depth_filter_maxdist"   value="5.0"/>
    <param name="occ_map/depth_filter_mindist"   value="0.2"/>
		<param name="occ_map/depth_filter_margin"    value="2"/>
    <param name="occ_map/skip_pixel" value="2"/>

    <!-- use when mapping frequency is high (>= 30Hz) -->
    <param name="occ_map/use_shift_filter" value="false" type="bool"/>
    <param name="occ_map/depth_filter_tolerance" value="0.15"/>
    
    <param name="occ_map/cx" value="321.04638671875"/>
    <param name="occ_map/cy" value="243.44969177246094"/>
    <param name="occ_map/fx" value="387.229248046875"/>
    <param name="occ_map/fy" value="387.229248046875"/>
    <param name="occ_map/rows" value="480"/>
    <param name="occ_map/cols" value="640"/>

    <param name="occ_map/prob_hit_log"  value="0.65"/>
    <param name="occ_map/prob_miss_log" value="0.0"/>
    <param name="occ_map/clamp_min_log"  value="0.12"/>
    <param name="occ_map/clamp_max_log"  value="0.90"/>
    <param name="occ_map/min_occupancy_log"  value="0.80"/>
    <param name="occ_map/min_ray_length" value="0.1"/>
    <param name="occ_map/max_ray_length" value="4.5"/>
    <param name="occ_map/use_global_map" value="false" type="bool"/>




    <param name="pos_checker/hrz_safe_radius" value="0.2"/>
    <param name="pos_checker/vtc_safe_radius" value="0.2"/>
    <param name="pos_checker/copter_diag_len" value="0.05"/>
    <param name="pos_checker/dt" value="0.015"/>
    <param name="pos_checker/inflate" value="false"/>
  
    <!-- kinorrt* params --> 
    <param name="krrt/rho" value="0.01" type="double"/> <!-- the quadratic matrix R of u'Ru -->
    <param name="krrt/vel_limit" value="7" type="double" />
    <param name="krrt/acc_limit" value="5" type="double" />
    <param name="krrt/tree_node_nums" value="5000" type="int" />
    <param name="krrt/orphan_nums" value="1500" type="int" />
    <param name="krrt/debug_vis" value="true" type="bool" />
    <param name="krrt/radius_cost_between_two_states" value="4" type="double" /> <!-- cost tolerance -->
    <param name="krrt/allow_orphan" value="false" type="bool" />
    <param name="krrt/allow_close_goal" value="false" type="bool" />
    <param name="krrt/stop_after_first_traj_found" value="true" type="bool" />
    <param name="krrt/rewire" value="false" type="bool" />

    <param name="sampler/vel_mag_mean" value="4.0" type="double" />
    <param name="sampler/pos_hrz_var" value="1.0" type="double" />
    <param name="sampler/pos_vtc_var" value="0.2" type="double" />
    <param name="sampler/vel_mag_var" value="0.7" type="double" />
    <param name="sampler/vel_dir_var" value="0.2" type="double" />
    <param name="sampler/resolution" value="0.1" type="double" />



  <!-- planner manager -->
    <param name="manager/max_vel" value="$(arg max_vel)" type="double"/>
    <param name="manager/max_acc" value="$(arg max_acc)" type="double"/>
    <param name="manager/max_jerk" value="4" type="double"/>
    <param name="manager/control_points_distance" value="0.4" type="double"/>
    <param name="manager/feasibility_tolerance" value="0.05" type="double"/>
    <param name="manager/planning_horizon" value="$(arg planning_horizon)" type="double"/>
    <param name="manager/use_distinctive_trajs" value="$(arg use_distinctive_trajs)" type="bool"/>
    <param name="manager/drone_id" value="$(arg drone_id)"/>







  <!-- trajectory optimization -->
   <param name="optimization/lambda_smooth_new" value="1" type="double"/>
   <param name="optimization/lambda_collision_new" value="0.05" type="double"/>
   <param name="optimization/lambda_feasibility_new" value="0.1" type="double"/>
    <param name="optimization/lambda_fitness_new" value="1.0" type="double"/>
    <param name="optimization/swarm_clearance_new" value="0.5" type="double"/>
    <param name="optimization/lambda_smooth" value="1" type="double"/> 
    <param name="optimization/lambda_collision" value="0.5" type="double"/>
    <param name="optimization/lambda_feasibility" value="0.1" type="double"/>
    <param name="optimization/lambda_fitness" value="0.6" type="double"/>
    



    <param name="optimization/lambda_smooth_re" value="1" type="double"/> 
    <param name="optimization/lambda_collision_re" value="0.8" type="double"/>
    <param name="optimization/lambda_feasibility_re" value="1" type="double"/>
    <param name="optimization/lambda_fitness_re" value="0.08" type="double"/>




    
    <param name="optimization/dist0" value="0.5" type="double"/>
    <param name="optimization/swarm_clearance" value="0.5" type="double"/>
    <param name="optimization/max_vel" value="$(arg max_vel)" type="double"/>
    <param name="optimization/max_acc" value="$(arg max_acc)" type="double"/>

    <param name="bspline/limit_vel" value="$(arg max_vel)" type="double"/>
    <param name="bspline/limit_acc" value="$(arg max_acc)" type="double"/>
    <param name="bspline/limit_ratio" value="1.1" type="double"/>

  <!-- objects prediction -->
    <param name="prediction/obj_num" value="$(arg obj_num_set)" type="int"/>
    <param name="prediction/lambda" value="1.0" type="double"/>
    <param name="prediction/predict_rate" value="1.0" type="double"/>
  
  


  </node>

</launch>
